cmd/snap-confine/snap-confine.c

```diff
@@ -623,6 +624,34 @@ static void enter_classic_execution_environment(const sc_invocation *inv,
 	}
 }
 
+static bool is_device_cgroup_self_managed(const sc_invocation *inv)
+{
+	char info_path[PATH_MAX] = { 0 };
+	sc_must_snprintf(info_path,
+			 sizeof info_path,
+			 "/var/lib/snapd/cgroup/snap.%s.device",
+			 inv->snap_instance);
+
+	FILE *stream SC_CLEANUP(sc_cleanup_file) = NULL;
+	stream = fopen(info_path, "r");
+	if (stream == NULL && errno == ENOENT) {
```

Comment:
we have a bit of the problem that with many other backends we fail if the backend profile is not there? but here we have a window in which we might run the snap set up in the wrong way if it is started too early?  maybe processing the udev backend before seccomp and apparmor would solve this?  see interfaces/backends/backends.go ?

Reply:
As discussed on MM, we have a bit of a problem here and a couple of possible solutions. In general, s-c does not synchronize the execution path with backend setup very well. In the s-c code, we actively wait for the seccomp profile, and we hope that the apparmor profile is present by the time we call `aa_change_on_exec()` (assuming apparmor wasn't disabled at built time). While the backend initialization is the following order:
https://github.com/snapcore/snapd/blob/eba5f39bb4ef2b7b5c6b9ff5845831d0e82a3462/interfaces/backends/backends.go#L38-L70 (with apparmor being optionally last). So there's a chance one could start an application before al the backends have their Setup() called, as there's no clear checkpoint. A couple of options:
- we can either add a null backend, whose only purpose is to provide a checkpoint file (where?, when is it cleaned up?)
- I can reorder the backends such that seccomp (shared by all build configurations), runs last and we implicitly (or rather explicitly this time), use the binary profile file as checkpoint
- I can make the `*.device` file mandatory, and thus we'd only have to wait for the udev backend to complete (ignoring the other backends, polkit, dbus, kmod, mount etc.)

Reply:
I've pushed changes to make the `*.device` file mandatory